numba/npyufunc/_internal.c

```diff
@@ -104,6 +105,557 @@ PyTypeObject PyUFuncCleaner_Type = {
+    }
+
+    if (!PyArg_ParseTuple(args, "kO!|k", &loop_ptr, &PyList_Type, &arg_types,
+                          &data_ptr)) {
+        return NULL;
+    }
+
+    arg_types_arr = _build_arg_types_array(arg_types, (Py_ssize_t)ufunc->nargs);
+    if (!arg_types_arr) goto _dufunc__add_loop_fail;
+    for (idx = 0; idx < ufunc->nargs; idx++) {
+        if (arg_types_arr[idx] >= NPY_USERDEF) {
+            usertype = arg_types_arr[idx];
+        }
+    }
+
+    if (usertype != NPY_VOID) {
+        if (PyUFunc_RegisterLoopForType(ufunc, usertype,
+                                        (PyUFuncGenericFunction)loop_ptr,
+                                        arg_types_arr, (void *)data_ptr) < 0) {
+            goto _dufunc__add_loop_fail;
+        }
+    } else if (PyUFunc_ReplaceLoopBySignature(ufunc,
+                                              (PyUFuncGenericFunction)loop_ptr,
+                                              arg_types_arr, &old_func) == 0) {
+        /* TODO: Consider freeing any memory held by the old loop (somehow) */
+        for (idx = 0; idx < ufunc->ntypes; idx++) {
+            if (ufunc->functions[idx] == (PyUFuncGenericFunction)loop_ptr) {
+                ufunc->data[idx] = (void *)data_ptr;
+                break;
+            }
+        }
+    } else {
+        int ntypes=ufunc->ntypes + 1;
+        PyUFuncGenericFunction *functions=NULL;
+        void **data=NULL;
+        char *types=NULL;
+        void *newptr=NULL, *oldptr=NULL;
+        size_t functions_size=sizeof(PyUFuncGenericFunction) * ntypes;
+        size_t data_size=sizeof(void *) * ntypes;
+        size_t type_ofs=sizeof(char) * ufunc->ntypes * ufunc->nargs;
+        size_t newsize=(functions_size + data_size +
+                        (sizeof(char) * ntypes * ufunc->nargs) +
+                        (sizeof(void *) - 1)); /* Alignment fudge... */
+
+        oldptr = ufunc->ptr;
+        newptr = PyArray_malloc(newsize);
+        functions =
+          (PyUFuncGenericFunction*)((char *)newptr +
+                                    ((size_t)newptr % sizeof(void *)));
+        memcpy(functions, ufunc->functions,
+               sizeof(PyUFuncGenericFunction) * ufunc->ntypes);
+        functions[ntypes - 1] = (PyUFuncGenericFunction)loop_ptr;
+        data = (void **)((char *)functions + functions_size);
+        memcpy(data, ufunc->data, sizeof(void *) * ufunc->ntypes);
+        data[ntypes - 1] = (void *)data_ptr;
+        types = (char *)data + data_size;
+        memcpy(types, ufunc->types, sizeof(char) * ufunc->ntypes *
+               ufunc->nargs);
+        for (idx = 0; idx < ufunc->nargs; idx++) {
+            types[idx + type_ofs] = (char)arg_types_arr[idx];
+        }
+
+        ufunc->ntypes = ntypes;
+        ufunc->functions = functions;
+        ufunc->types = types;
+        ufunc->data = data;
+        ufunc->ptr = newptr;
+        PyArray_free(oldptr);
+    }
+
+    PyArray_free(arg_types_arr);
+    Py_INCREF(Py_None);
```

Comment:
I don't like it.  It's bad enough that this code is not single-entry-single-exit.